---
title: "QualityControl"
author: "Xiang Ji"
date: "11/16/2017"
output: html_document
---

This file summarizes the data sets in Harpak et al. 2017. Main interests are 
length of alignment, level of missing species etc.

```{r}
rm(list=ls())  # clean up workspace
setwd("/Users/xji3/GitFolders/YeastIGCTract/Harpak/")
#setwd("/Users/Xiang/GitFolders/YeastIGCTract/Harpak/")
alignment.files <- list.files("./intronAlignmentSeparated/")
filtered.alignment.files <- NULL
num.missing.genes.list <- NULL
filtered.files <- NULL
for(seq.file in alignment.files){
  info <- read.table(paste("./intronAlignmentSeparated/", seq.file, sep = ""), as.is = TRUE)
  num.missing.genes <- sum(is.na(info[, 1]))
  num.missing.genes.list <- c(num.missing.genes.list, num.missing.genes)
  if(num.missing.genes == 0){
    filtered.files <- c(filtered.files, seq.file)
  }
  if(dim(info)[2] > 1){
    filtered.alignment.files <- c(filtered.alignment.files, seq.file)
  }
}
table(num.missing.genes.list)
```

OK, 27 datasets have no missing genes (based on gene names). Now apply additional filter by looking at the gap percentage (100% gaps = missing).

```{r}
additional.failure <- NULL
for(seq.file in filtered.files){
  info <- read.table(paste("./intronAlignmentSeparated/", seq.file, sep = ""), as.is = TRUE)
  for(row.num in 1:11){
    seq <- strsplit(info[row.num, 2], "")[[1]]
    percent.gap <- (length(grep("-", seq)) + length(grep("N", seq)))/length(seq)
    if (percent.gap > 1 - 1e-6){
      if( ! seq.file %in% additional.failure)
        additional.failure <- c(additional.failure, seq.file)
    }
  }
}
additional.filtered <- filtered.files[! filtered.files %in% additional.failure]
# See how many have missing info
length(additional.failure)
# See how many have passed the additional filter
length(additional.filtered)
# Show the remaining data sets
additional.filtered
```

Now summarize all alignment files in terms of how many ingroup species have missing genes.

```{r}
num.missing.species.list  <- NULL
for(num.spe in 0:5){
  assign(paste("missing", num.spe, "species", sep = "."), NULL)
}
for(seq.file in filtered.alignment.files){
  info <- read.table(paste("./intronAlignmentSeparated/", seq.file, sep = ""), as.is = TRUE)
  for(row.num in 1:11){
    seq <- strsplit(info[row.num, 2], "")[[1]]
    percent.gap <- length(grep("-", seq))/length(seq)
    if (percent.gap > 1 - 1e-6){
      info[row.num, 1] <- NA
    }   
  }
  num.spe <- 0
  for(row.iter in 1:5){
    if(any(is.na(info[c(row.iter*2-1, row.iter*2), 1]))){
      num.spe <- num.spe + 1
    }
  }
  num.missing.species.list <- c(num.missing.species.list, num.spe)
  assign(paste("missing", num.spe, "species", sep = "."), c(get(paste("missing", num.spe, "species", sep = ".")), seq.file))
}
table(num.missing.species.list)
# In addition, show files that have only one ingroup species that misses either one or both copies
missing.1.species
```

Now look at the length of the alignment and percent of columns that contain gaps

```{r}
# 20 files that have no missing genes
missing.0.summary <- NULL
for(seq.file in missing.0.species){
  info <- read.table(paste("./intronAlignmentSeparated/", seq.file, sep = ""), as.is = TRUE)

  gap.col <- NULL
  for(row.num in 1:11){
    seq <- strsplit(info[row.num, 2], "")[[1]]
    gap.col <- union(gap.col, grep("-", seq))
  }
  
  missing.0.summary <- rbind(missing.0.summary, c(length(seq), length(seq) - length(gap.col), length(gap.col), length(gap.col)/length(seq)))
}
rownames(missing.0.summary) <- missing.0.species
colnames(missing.0.summary) <- c("length", "# col no gap", "# col with gap", "% gap col")
missing.0.summary

# 46 files that have one species that contains missing genes
missing.1.summary <- NULL
for(seq.file in missing.1.species){
  info <- read.table(paste("./intronAlignmentSeparated/", seq.file, sep = ""), as.is = TRUE)

  gap.col <- NULL
  for(row.num in 1:11){
    seq <- strsplit(info[row.num, 2], "")[[1]]
    if(length(grep("-", seq)) < length(seq)){
      gap.col <- union(gap.col, grep("-", seq))
    }
  }
  
  missing.1.summary <- rbind(missing.1.summary, c(length(seq), length(seq) - length(gap.col), length(gap.col), length(gap.col)/length(seq)))
}
rownames(missing.1.summary) <- missing.1.species
colnames(missing.1.summary) <- c("length", "# col no gap", "# col with gap", "% gap col")
missing.1.summary

# Now save the file names
write.table(missing.0.species, "./missing_0_species_list.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
write.table(missing.1.species, "./missing_1_species_list.txt", row.names = FALSE, col.names = FALSE, quote = FALSE)
```

####01042018 update

```{r}
# count groups
group.list <- NULL
for(name in filtered.alignment.files){
  group.name <- strsplit(name, "\\.")[[1]][1]
  group.list <- c(group.list, group.name)
}
unique.groups <- unique(group.list)
length(unique.groups)
unique.groups

missing.0.groups <- NULL
for(name in missing.0.species){
  group.name <- strsplit(name, "\\.")[[1]][1]
  missing.0.groups <- c(missing.0.groups, group.name)
}
missing.0.groups <- unique(missing.0.groups)
length(missing.0.groups)
missing.0.groups

# Now add up total number of sites analyzed
total.sites <- 0
for(seq.file in missing.0.species){
  info <- read.table(paste("./intronAlignmentSeparated/", seq.file, sep = ""), as.is = TRUE)
  total.sites <- total.sites + nchar(info[1, 2])
}
total.sites
```


####01172018 update

```{r}
# Count how many files that contain species with only one paralog
ghost.lineage.files <- NULL
for(seq.file in filtered.alignment.files){
  info <- read.table(paste("./intronAlignmentSeparated/", seq.file, sep = ""), as.is = TRUE)
  for(row.num in 1:11){
    seq <- strsplit(info[row.num, 2], "")[[1]]
    percent.gap <- length(grep("-", seq))/length(seq)
    if (percent.gap > 1 - 1e-6){
      info[row.num, 1] <- NA
    }   
  }
  for(row.iter in 1:5){
    if(sum(is.na(info[c(row.iter*2-1, row.iter*2), 1])) == 1){
      ghost.lineage.files <- c(ghost.lineage.files, seq.file)
    }
  }
}
ghost.lineage.files <- unique(ghost.lineage.files)
# number of alignment files that contain ghost lineage
length(ghost.lineage.files)

ghost.groups <- NULL
for(name in ghost.lineage.files){
  group.name <- strsplit(name, "\\.")[[1]][1]
  ghost.groups <- c(ghost.groups, group.name)
}
ghost.groups <- unique(ghost.groups)
# number of gene pairs that contain ghost lineage
length(ghost.groups)
```

####01182018 update

Got human segmental duplication file from : http://humanparalogy.gs.washington.edu/build38/data/GRCh38GenomicSuperDup.tab

Ensembl ID | Chrm | name
------------- | ------------- |  ------------- 
[ENSG00000109272](http://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000109272;r=4:73853189-73854155;t=ENST00000226524) | 4 | PF4V1
[ENSG00000163737](http://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163737;r=4:73981077-73982124;t=ENST00000296029) | 4 | PF4
[ENSG00000136943](http://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000136943;r=9:97029679-97039643) | 9 | CTSV
[ENSG00000135047](http://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000135047;r=9:87725519-87731393) | 9 | CTSL
[ENSG00000158485](http://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000158485;r=1:158327951-158331531) | 1 | CD1B
[ENSG00000158477](https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000158477;r=1:158254137-158258269;t=ENST00000289429) | 1 | CD1A
[ENSG00000163564](http://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163564;r=1:158930796-158977054) | 1 | PYHIN1
[ENSG00000163563](https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000163563;r=1:158831317-158849506) | 1 | PYHIN3
[ENSG00000187626](http://uswest.ensembl.org/Homo_sapiens/Gene/Summary?db=core;g=ENSG00000187626;r=6:28244623-28252224;t=ENST00000377294) | 6 | ZKSCAN4
[ENSG00000189298](https://uswest.ensembl.org/Homo_sapiens/Gene/Summary?g=ENSG00000189298;r=6:28349914-28369177) | 6 |ZKSCAN3

After all, I give up on checking the duplication regions.  It's pretty hard to locate even one gene into the start and end point that Eichler's lab provided. 
I don't know how the pairs were constructed.

```{r}
sg.dup.data <- read.table("~/Downloads/GRCh38GenomicSuperDup.tab", header = FALSE, 
                          stringsAsFactors = FALSE)
dup.header <- readLines("~/Downloads/GRCh37GenomicSuperDup.tab", n=1)
colnames(sg.dup.data) <- strsplit(dup.header, "\t")[[1]]
boundaries <- c(158327951, 158331531, 158254137, 158258269, 158930796, 158977054, 158831317, 158849506)
boundaries.1.1 <- c(158327951, 158331531)
boundaries.1.2 <- c(158254137, 158258269)
boundaries.2.1 <- c(158930796, 158977054)
boundaries.2.2 <- c(158831317, 158849506)

same.region <- FALSE
pair.1.region <- FALSE
pair.2.region <- FALSE
start.list <- NULL
end.list <- NULL
store.row.num.list <- NULL
for(row.num in 1:dim(sg.dup.data)[1]){
  if (substring(sg.dup.data[row.num, "chrom"], 4) == "1" ){
      # & substring(sg.dup.data[row.num, "otherChrom"], 4) == "1"){
    sg.start.1 <- sg.dup.data[row.num, "chromStart"]
    sg.end.1 <- sg.dup.data[row.num, "chromEnd"]
    sg.start.2 <- sg.dup.data[row.num, "otherStart"]
    sg.end.2 <- sg.dup.data[row.num, "otherEnd"]
    
    sg.start <- min(c(sg.start.1, sg.start.2))
    sg.end <- max(c(sg.end.1, sg.end.2))
    
    if(sg.start < min(boundaries) & sg.end > max(boundaries)){
      same.region <- TRUE
      store.row.num.list <- c(store.row.num.list, row.num)
    }
    if(sg.start < min(boundaries.1.1) & sg.end > max(boundaries.1.1)){
      pair.1.region <- TRUE
    }

    if(sg.start < min(boundaries.1.2) & max(boundaries.1.2) < sg.end){
      pair.1.region <- TRUE
    }

    if((sg.start.1 < min(boundaries.2.1) & sg.end.1 > min(boundaries.2.1)) 
       | (sg.start.2 < min(boundaries.2.1) & sg.end.2 > min(boundaries.2.1))){
      pair.2.region <- TRUE
    }

    # if(sg.start < min(boundaries.2.2) & max(boundaries.2.2) < sg.end){
    #   pair.2.region <- TRUE
    # }    
    # 
    start.list <- c(start.list, sg.dup.data[row.num, "chromStart"])
    end.list <- c(end.list, sg.dup.data[row.num, "chromEnd"])
  }
}


```


#### 01212018 update

Check if Harpak et al removed 50bp flanking region of the introns.

```{r}
library(biomaRt)
Ensembl.id.list <- c("ENSG00000109272", "ENSG00000163737", "ENSG00000136943", "ENSG00000135047", 
                     "ENSG00000158485","ENSG00000158477", "ENSG00000163564", "ENSG00000163563", "ENSG00000187626")
#                     "ENSG0000018929")  This one has trouble using getSequence()
group.id.list <- c()
mart <- useMart("ensembl", dataset="hsapiens_gene_ensembl")

for(Ensembl.id in Ensembl.id.list){
  exon.seq <- getSequence(id = Ensembl.id, type = "ensembl_gene_id", seqType = "gene_exon", mart = mart)
  exon.intron.seq <- getSequence(id = Ensembl.id[1], type = "ensembl_gene_id", seqType = "gene_exon_intron", mart = mart)
  
  exon.pos.list <- NULL
  for(exon in exon.seq$gene_exon){
    exon.pos.list <- c(exon.pos.list, gregexpr(exon, exon.intron.seq$gene_exon_intron[1], perl = TRUE)[[1]][1])
  }
  intron.length.list <- NULL
  intron.list <- NULL
  left.seq <- exon.intron.seq$gene_exon_intron[1]
  for(exon in exon.seq$gene_exon[order(exon.pos.list)]){
    new.split <- strsplit(left.seq, exon, perl = TRUE)[[1]]
    if(length(new.split) == 1){
      left.seq <- new.split[1]
      intron.length.list <- c(intron.length.list, nchar(new.split[1]))
      intron.list <- c(intron.list, new.split[1])
    }else{
      if(nchar(new.split[1])){
        intron.length.list <- c(intron.length.list, nchar(new.split[1]))
        intron.list <- c(intron.list, new.split[1])
      }
      left.seq <- new.split[2]
    }
    
  }
  cat("Ensembl ID: ", Ensembl.id, "\n", 
      "Intron lengths: ", intron.length.list, "\n", 
      "# Intron length > 98: ", sum(intron.length.list > 98), "\n")
}
```